src/spi.rs

```diff
@@ -414,14 +437,58 @@ impl<'d> SpiMasterDriver<'d> {
             quadwp_io_num: -1,
             quadhd_io_num: -1,
 
-            max_transfer_sz: config.dma.max_transfer_size() as i32,
+            max_transfer_sz: max_transfer_sz as i32,
             ..Default::default()
         };
 
-        esp!(unsafe { spi_bus_initialize(SPI::device(), &bus_config, config.dma.into()) })?;
+        esp!(unsafe { spi_bus_initialize(SPI::device(), &bus_config, dma_chan) })?;
+        Ok(max_transfer_sz)
+    }
+}
+
+impl<'d> Drop for SpiMasterDriver<'d> {
+    fn drop(&mut self) {
+        // should we drop without checking the lock?
+        let guard = self.lock.enter();
+        esp!(unsafe { spi_bus_free(self.handle) }).unwrap();
+    }
+}
+
+pub struct EspSpiDevice<'d, T> {
+    handle: Option<spi_device_handle_t>,
+    cs_pin_ref: PeripheralRef<'d, AnyOutputPin>,
+    config: spi_device_interface_config_t,
+    driver: T,
+    _p: PhantomData<&'d ()>,
+}
+
+impl<'d, T> EspSpiDevice<'d, T>
+where
+    T: Borrow<SpiMasterDriver<'d>> + 'd,
+{
+    pub fn new(
+        driver: T,
+        cs: impl Peripheral<P = impl InputPin + OutputPin> + 'd,
+        config: config::Config,
```

Comment:
For consistency with other drivers.
```suggestion
        config: &config::Config,
```